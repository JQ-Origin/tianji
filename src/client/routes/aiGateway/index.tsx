import { CommonPageEmpty } from '@/components/CommonPageEmpty';
import { routeAuthBeforeLoad } from '@/utils/route';
import { useTranslation } from '@i18next-toolkit/react';
import { createFileRoute } from '@tanstack/react-router';

export const Route = createFileRoute('/aiGateway/')({
  beforeLoad: routeAuthBeforeLoad,
  component: PageComponent,
});

function PageComponent() {
  const { t } = useTranslation();

  return (
    <CommonPageEmpty
      text={t('Not any ai gateway has been exist')}
      buttonText={t('Add Ai Gateway Now')}
      buttonUrl="/aiGateway/add"
    />
  );
}
