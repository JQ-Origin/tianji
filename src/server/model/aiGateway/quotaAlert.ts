import { prisma } from '../_client.js';
import { sendNotification } from '../notification/index.js';
import { token } from '../notification/token/index.js';
import { logger } from '../../utils/logger.js';
import { buildQueryWithCache, getCacheManager } from '../../cache/index.js';
import { withDistributedLock } from '../../cache/distributedLock.js';
import dayjs from 'dayjs';

/**
 * Generate daily cost cache key
 */
function getDailyCostCacheKey(
  workspaceId: string,
  gatewayId: string,
  date?: string | dayjs.Dayjs
): string {
  const dateStr = date
    ? dayjs(date).format('YYYY-MM-DD')
    : dayjs().format('YYYY-MM-DD');
  return `ai-gateway-daily-cost:${workspaceId}:${gatewayId}:${dateStr}`;
}

export interface QuotaAlertLevel {
  percentage: number;
  level: '80' | '100' | '150';
  field: 'alertLevel80Sent' | 'alertLevel100Sent' | 'alertLevel150Sent';
  title: string;
  color: string;
}

export const QUOTA_ALERT_LEVELS: QuotaAlertLevel[] = [
  {
    percentage: 80,
    level: '80',
    field: 'alertLevel80Sent',
    title: 'AI Gateway Quota Alert - 80%',
    color: 'yellow',
  },
  {
    percentage: 100,
    level: '100',
    field: 'alertLevel100Sent',
    title: 'AI Gateway Quota Alert - 100%',
    color: 'orange',
  },
  {
    percentage: 150,
    level: '150',
    field: 'alertLevel150Sent',
    title: 'AI Gateway Quota Alert - 150%',
    color: 'red',
  },
];

// Cache for quota alert settings (5 minutes TTL)
const {
  get: getQuotaAlertCache,
  del: clearQuotaAlertCache,
  update: updateQuotaAlertCache,
} = buildQueryWithCache(async (workspaceId: string, gatewayId: string) => {
  return await prisma.aIGatewayQuotaAlert.findFirst({
    where: {
      workspaceId,
      gatewayId,
      enabled: true,
    },
    include: {
      aiGateway: true,
      notification: true,
    },
  });
});

/**
 * Clear quota alert cache for a specific gateway
 */
export function clearQuotaAlertCacheForGateway(
  workspaceId: string,
  gatewayId: string
) {
  clearQuotaAlertCache(workspaceId, gatewayId);
}

/**
 * Get daily cost from cache or calculate if not cached
 */
async function getDailyCost(
  workspaceId: string,
  gatewayId: string
): Promise<number> {
  const cacheKey = getDailyCostCacheKey(workspaceId, gatewayId);
  const cacheManager = await getCacheManager();

  // Try to get from cache first
  const cachedCost = await cacheManager.get(cacheKey);
  if (cachedCost !== undefined) {
    return Number(cachedCost);
  }

  // Calculate fresh value
  const todayStart = dayjs().startOf('day');
  const tomorrowStart = dayjs().add(1, 'day').startOf('day');

  const todayCost = await prisma.aIGatewayLogs.aggregate({
    where: {
      gatewayId,
      workspaceId,
      status: 'Success',
      createdAt: {
        gte: todayStart.toDate(),
        lt: tomorrowStart.toDate(),
      },
    },
    _sum: {
      price: true,
    },
  });

  const totalCost = Number(todayCost._sum.price || 0);

  // Cache for 5 minutes
  await cacheManager.set(cacheKey, totalCost, 5 * 60 * 1000);

  return totalCost;
}

/**
 * Update daily cost cache after a new request
 */
async function updateDailyCostCache(
  workspaceId: string,
  gatewayId: string,
  additionalCost: number
): Promise<void> {
  const cacheKey = getDailyCostCacheKey(workspaceId, gatewayId);
  const cacheManager = await getCacheManager();

  // Get current cached value
  const cachedCost = await cacheManager.get(cacheKey);
  if (cachedCost !== undefined) {
    // Update cache with new total
    const newTotal = Number(cachedCost) + additionalCost;
    await cacheManager.set(cacheKey, newTotal, 5 * 60 * 1000);
  }
  // If not cached, it will be calculated fresh next time
}

/**
 * Check and send quota alerts for a specific gateway
 * This function should be called after each AI Gateway request
 */
export async function checkQuotaAlert(
  workspaceId: string,
  gatewayId: string,
  requestCost?: number
): Promise<void> {
  try {
    // Get quota alert settings from cache
    const quotaAlert = await getQuotaAlertCache(workspaceId, gatewayId);

    if (
      !quotaAlert ||
      !quotaAlert.notificationId ||
      Number(quotaAlert.dailyQuota) <= 0
    ) {
      return;
    }

    // Update daily cost cache if requestCost is provided
    if (requestCost !== undefined && requestCost > 0) {
      await updateDailyCostCache(workspaceId, gatewayId, requestCost);
    }

    // Get today's total cost from cache
    const totalCost = await getDailyCost(workspaceId, gatewayId);
    const dailyQuota = Number(quotaAlert.dailyQuota);
    const percentage = (totalCost / dailyQuota) * 100;

    logger.info(
      `[checkQuotaAlert] Gateway ${gatewayId}: cost=${totalCost}, quota=${dailyQuota}, percentage=${percentage.toFixed(1)}%`
    );

    // Check each alert level
    for (const alertLevel of QUOTA_ALERT_LEVELS) {
      const shouldAlert = percentage >= alertLevel.percentage;
      const alreadySent = quotaAlert[alertLevel.field];

      if (shouldAlert && !alreadySent) {
        await withDistributedLock(
          `quota-alert-notification:${workspaceId}:${gatewayId}:${alertLevel.level}`,
          async () => {
            logger.info(
              `[checkQuotaAlert] Sending ${alertLevel.level}% alert for gateway ${gatewayId}`
            );

            // Send notification
            await sendNotification(
              {
                name: quotaAlert.notification?.name ?? '',
                type: quotaAlert.notification?.type ?? '',
                payload: quotaAlert.notification?.payload ?? {},
              },
              `${alertLevel.title} - ${quotaAlert.aiGateway.name}`,
              [
                token.paragraph(`Gateway: ${quotaAlert.aiGateway.name}`),
                token.paragraph(`Today's Usage Cost: $${totalCost.toFixed(4)}`),
                token.paragraph(`Quota Set: $${dailyQuota.toFixed(4)}`),
                token.paragraph(`Usage Percentage: ${percentage.toFixed(1)}%`),
                token.paragraph(
                  `Alert Time: ${dayjs().format('YYYY-MM-DD HH:mm:ss')}`
                ),
                token.newline(),
                token.text(getAlertMessage(alertLevel.level, percentage)),
              ],
              {
                gatewayId: quotaAlert.gatewayId,
                gatewayName: quotaAlert.aiGateway.name,
                dailyCost: totalCost,
                dailyQuota,
                percentage,
                alertLevel: alertLevel.level,
                alertTime: dayjs().toISOString(),
              }
            );

            // local update quota alert cache to avoid high concurrency issue
            await updateQuotaAlertCache(
              workspaceId,
              gatewayId
            )({
              ...quotaAlert,
              [alertLevel.field]: true,
              lastAlertSentAt: new Date(),
            });

            // Update alert sent status
            await prisma.aIGatewayQuotaAlert.update({
              where: {
                id: quotaAlert.id,
              },
              data: {
                [alertLevel.field]: true,
                lastAlertSentAt: new Date(),
              },
            });

            // Clear cache after updating alert status
            clearQuotaAlertCache(workspaceId, gatewayId);

            logger.info(
              `[checkQuotaAlert] ${alertLevel.level}% alert sent for gateway ${gatewayId}`
            );
          },
          {
            timeout: 30_000, // 30 seconds timeout
            skipOnFailure: true, // Skip if another process is already sending the alert
          }
        );
      }
    }
  } catch (error) {
    logger.error(
      `[checkQuotaAlert] Error checking quota alert for gateway ${gatewayId}:`,
      error
    );
  }
}

/**
 * Reset daily alert flags (should be called at the start of each day)
 */
export async function resetDailyAlertFlags(): Promise<void> {
  try {
    // Get all enabled quota alerts to clear their caches
    const quotaAlerts = await prisma.aIGatewayQuotaAlert.findMany({
      where: {
        enabled: true,
      },
      select: {
        workspaceId: true,
        gatewayId: true,
      },
    });

    // Reset alert flags
    await prisma.aIGatewayQuotaAlert.updateMany({
      where: {
        enabled: true,
      },
      data: {
        alertLevel80Sent: false,
        alertLevel100Sent: false,
        alertLevel150Sent: false,
      },
    });

    // Clear quota alert caches
    for (const alert of quotaAlerts) {
      clearQuotaAlertCache(alert.workspaceId, alert.gatewayId);
    }

    // Clear daily cost caches for yesterday and today
    const cacheManager = await getCacheManager();
    const today = dayjs();
    const yesterday = dayjs().subtract(1, 'day');

    for (const alert of quotaAlerts) {
      const todayCacheKey = getDailyCostCacheKey(
        alert.workspaceId,
        alert.gatewayId,
        today
      );
      const yesterdayCacheKey = getDailyCostCacheKey(
        alert.workspaceId,
        alert.gatewayId,
        yesterday
      );

      await cacheManager.delete(todayCacheKey);
      await cacheManager.delete(yesterdayCacheKey);
    }

    logger.info(
      '[resetDailyAlertFlags] Daily alert flags reset and caches cleared'
    );
  } catch (error) {
    logger.error('[resetDailyAlertFlags] Error resetting alert flags:', error);
  }
}

// Returns an English alert message for AI Gateway quota usage based on the alert level
function getAlertMessage(level: string, percentage: number): string {
  switch (level) {
    case '80':
      return '‚ö†Ô∏è Your AI Gateway usage has reached 80% of the quota. Please monitor your usage.';
    case '100':
      return 'üö® Your AI Gateway usage has reached 100% of the quota. Please check your usage immediately.';
    case '150':
      return 'üî• Your AI Gateway usage has exceeded 150% of the quota. Please take immediate action to control costs!';
    default:
      return 'Please check your AI Gateway usage in time to avoid exceeding your budget.';
  }
}
