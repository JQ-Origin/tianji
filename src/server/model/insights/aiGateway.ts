import { z } from 'zod';
import { insightsQuerySchema } from '../../utils/schema.js';
import { prisma } from '../_client.js';
import { printSQL } from '../../utils/prisma.js';
import { Prisma } from '@prisma/client';
import { FilterInfoType, FilterInfoValue, getDateArray } from '@tianji/shared';
import { get, uniq } from 'lodash-es';
import { env } from '../../utils/env.js';
import { InsightsSqlBuilder } from './shared.js';

export class AIGatewayInsightsSqlBuilder extends InsightsSqlBuilder {
  getTableName() {
    return 'AIGatewayLogs';
  }

  buildSelectQueryArr() {
    const { metrics } = this.query;
    return metrics.map((item) => {
      if (item.math === 'events') {
        if (item.name === '$all_event') {
          return Prisma.sql`count(1) as "$all_event"`;
        }

        // For standard fields, directly count
        // TODO: maybe use other math functions
        if (['inputToken', 'outputToken', 'price'].includes(item.name)) {
          return Prisma.sql`sum("AIGatewayLogs"."${Prisma.raw(item.name)}") as ${Prisma.raw(`"${item.name}"`)}`;
        }
      } else if (item.math === 'sessions') {
        // AIGatewayLogs has no concept of sessions, but can be grouped by gatewayId
        if (item.name === '$all_event') {
          return Prisma.sql`count(distinct "gatewayId") as "$all_event"`;
        }
      }

      return null;
    });
  }

  buildWhereQueryArr() {
    const { insightId, time, filters } = this.query;
    const { startAt, endAt } = time;

    return [
      // gatewayId
      Prisma.sql`"AIGatewayLogs"."gatewayId" = ${insightId}`,

      // date range
      this.buildDateRangeQuery('"AIGatewayLogs"."createdAt"', startAt, endAt),

      ...filters.map((filter) =>
        this.buildFilterQueryOperator(
          filter.name,
          filter.type,
          filter.operator,
          filter.value
        )
      ),
    ];
  }

  private buildFilterQueryOperator(
    name: string,
    type: FilterInfoType,
    operator: string,
    value: FilterInfoValue | null
  ): Prisma.Sql {
    // Process standard fields
    if (
      [
        'gatewayId',
        'modelName',
        'status',
        'stream',
        'inputToken',
        'outputToken',
        'duration',
        'ttft',
        'price',
      ].includes(name)
    ) {
      const valueField = Prisma.sql`"AIGatewayLogs"."${Prisma.raw(name)}"${type === 'number' ? Prisma.empty : type === 'boolean' ? Prisma.empty : Prisma.empty}`;
      return this.buildCommonFilterQueryOperator(
        type,
        operator,
        value,
        valueField
      );
    }
    // Process JSON fields
    else if (name.startsWith('request.')) {
      const field = name.replace('request.', '');
      const valueField = Prisma.sql`("AIGatewayLogs"."requestPayload"->> '${field}')${type === 'number' ? Prisma.raw('::int') : type === 'boolean' ? Prisma.raw('::boolean') : Prisma.empty}`;
      return this.buildCommonFilterQueryOperator(
        type,
        operator,
        value,
        valueField
      );
    } else if (name.startsWith('response.')) {
      const field = name.replace('response.', '');
      const valueField = Prisma.sql`("AIGatewayLogs"."responsePayload"->> '${field}')${type === 'number' ? Prisma.raw('::int') : type === 'boolean' ? Prisma.raw('::boolean') : Prisma.empty}`;
      return this.buildCommonFilterQueryOperator(
        type,
        operator,
        value,
        valueField
      );
    }

    return Prisma.sql`1 = 1`;
  }
}

export async function insightsAIGateway(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string }
) {
  const { time, metrics, groups } = query;
  const { startAt, endAt, unit, timezone = context.timezone } = time;

  const builder = new AIGatewayInsightsSqlBuilder(query, context);
  const sql = builder.build();

  if (env.isDev) {
    printSQL(sql);
  }

  const res = await prisma.$queryRaw<{ date: string | null }[]>(sql);

  let result: {
    name: string;
    [groupName: string]: any;
    data: {
      date: string;
      value: number;
    }[];
  }[] = [];

  for (const m of metrics) {
    if (groups.length > 0) {
      for (const g of groups) {
        const allGroupValue = uniq(
          res.map((item) => get(item, `%${g.value}`) as any)
        );

        result.push(
          ...allGroupValue.map((gv) => ({
            name: m.name,
            [g.value]: gv,
            data: getDateArray(
              res
                .filter((item) => get(item, `%${g.value}`) === gv)
                .map((item) => {
                  return {
                    value: Number(get(item, m.name)),
                    date: String(item.date),
                  };
                }),
              startAt,
              endAt,
              unit
              // timezone // NOTICE: because return date is already with user timezone, so we don't need to pass timezone here
            ),
          }))
        );
      }
    } else {
      result.push({
        name: m.name,
        data: getDateArray(
          res.map((item) => {
            return {
              value: Number(get(item, m.name)),
              date: String(item.date),
            };
          }),
          startAt,
          endAt,
          unit
          // timezone // NOTICE: because return date is already with user timezone, so we don't need to pass timezone here
        ),
      });
    }
  }

  return result;
}
