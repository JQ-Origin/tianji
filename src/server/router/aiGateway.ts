import { type RequestHandler, Router } from 'express';
import { calcOpenAIToken } from '../model/openai.js';
import { z } from 'zod';
import OpenAI from 'openai';
import { prisma } from '../model/_client.js';
import { AIGatewayLogs, AIGatewayLogsStatus } from '@prisma/client';
import { getLLMCostDecimal } from '../utils/llm.js';
import { get } from 'lodash-es';

export const aiGatewayRouter = Router();

const openaiRequestSchema = z
  .object({
    model: z.string(),
    messages: z.array(z.any()).nonempty(),
    temperature: z.number().optional(),
    top_p: z.number().optional(),
    max_tokens: z.number().int().optional(),
    stream: z.boolean().optional(),
    presence_penalty: z.number().optional(),
    frequency_penalty: z.number().optional(),
  })
  .passthrough();

aiGatewayRouter.post(
  '/v1/:workspaceId/:gatewayId/openai/chat/completions',
  buildOpenAIHandler({})
);

aiGatewayRouter.post(
  '/v1/:workspaceId/:gatewayId/deepseek/chat/completions',
  buildOpenAIHandler({
    baseUrl: 'https://api.deepseek.com',
    modelPriceName: (model) => {
      if (model.startsWith('deepseek/')) {
        return model;
      }

      return `deepseek/${model}`;
    },
  })
);

aiGatewayRouter.post(
  '/v1/:workspaceId/:gatewayId/openrouter/chat/completions',
  buildOpenAIHandler({
    baseUrl: 'https://openrouter.ai/api/v1',
    modelPriceName: (model) => {
      if (model.startsWith('openrouter/')) {
        return model;
      }

      if (model.includes('deepseek-chat')) {
        return `openrouter/deepseek/deepseek-chat`;
      }

      if (model.includes('deepseek-coder')) {
        return `openrouter/deepseek/deepseek-coder`;
      }

      if (model.includes('deepseek-r1')) {
        return `openrouter/deepseek/deepseek-r1`;
      }

      return `openrouter/${model}`;
    },
  })
);

interface OpenaiHandlerOptions {
  baseUrl?: string;
  modelPriceName?: (model: string) => string;
}

export function buildOpenAIHandler(
  options: OpenaiHandlerOptions
): RequestHandler {
  return async (req, res) => {
    const payload = openaiRequestSchema.parse(req.body);
    const { model, messages, stream } = payload;
    const { workspaceId, gatewayId } = z
      .object({
        workspaceId: z.string(),
        gatewayId: z.string(),
      })
      .parse(req.params);
    const apiKey = (req.headers.authorization ?? '').replace('Bearer ', '');
    const start = Date.now();

    const logP = new Promise<AIGatewayLogs>(async (resolve) => {
      const inputToken = messages.reduce((acc, msg) => {
        return acc + calcOpenAIToken(String(msg.content), model);
      }, 0);

      const _log = await prisma.aIGatewayLogs.create({
        data: {
          workspaceId,
          gatewayId,
          modelName: model,
          stream,
          inputToken,
          outputToken: -1,
          duration: -1,
          ttft: -1,
          requestPayload: payload,
          responsePayload: {},
          status: AIGatewayLogsStatus.Pending,
        },
      });

      resolve(_log);
    });

    try {
      const openai = new OpenAI({
        apiKey,
        baseURL: options.baseUrl,
      });
      const modelPriceName = options.modelPriceName
        ? options.modelPriceName(model)
        : model;

      // Handle stream response
      if (stream) {
        res.setHeader('Content-Type', 'text/event-stream');
        res.setHeader('Cache-Control', 'no-cache');
        res.setHeader('Connection', 'keep-alive');

        const stream = await openai.chat.completions.create({
          ...payload,
          stream: true,
        });

        let outputContent = '';
        let ttft = -1;
        let modelName = model; // real model name which returned from remote. its will be some changed because of the model alias.
        for await (const chunk of stream) {
          modelName = chunk.model;
          if (ttft === -1) {
            ttft = Date.now() - start;
          }
          const content = get(chunk, ['choices', 0, 'delta', 'content']) || '';
          outputContent += content;
          res.write(`data: ${JSON.stringify(chunk)}\n\n`);

          if (res.flush) {
            res.flush();
          }
        }

        res.write('data: [DONE]\n\n');
        res.end();
        const duration = Date.now() - start;

        logP.then(async ({ id: logId, inputToken }) => {
          const outputToken = calcOpenAIToken(outputContent, model);

          await prisma.aIGatewayLogs.update({
            where: {
              id: logId,
            },
            data: {
              status: AIGatewayLogsStatus.Success,
              modelName,
              outputToken,
              duration,
              ttft,
              price: getLLMCostDecimal(modelPriceName, inputToken, outputToken),
              responsePayload: { content: outputContent },
            },
          });
        });
      } else {
        // Handle normal response
        const response = await openai.chat.completions.create({
          ...payload,
          stream: false,
        });

        const modelName = response.model ?? model;

        res.json(response);
        const duration = Date.now() - start;

        logP.then(async ({ id: logId, inputToken }) => {
          const content = get(response, ['choices', 0, 'message', 'content']);

          const outputToken =
            response.usage?.completion_tokens ??
            (typeof content === 'string' ? calcOpenAIToken(content, model) : 0);

          await prisma.aIGatewayLogs.update({
            where: {
              id: logId,
            },
            data: {
              status: AIGatewayLogsStatus.Success,
              outputToken,
              duration,
              modelName,
              price: getLLMCostDecimal(modelPriceName, inputToken, outputToken),
              responsePayload: { ...response },
            },
          });
        });
      }
    } catch (error) {
      // Handle API error
      console.error('OpenAI API error:', error);
      res.status(500).json({
        error: {
          message: error instanceof Error ? error.message : 'Unknown error',
          type: 'server_error',
        },
      });

      const duration = Date.now() - start;
      logP.then(async ({ id: logId }) => {
        await prisma.aIGatewayLogs.update({
          where: {
            id: logId,
          },
          data: {
            status: AIGatewayLogsStatus.Failed,
            duration,
            responsePayload: {
              error: String(error),
            },
          },
        });
      });
    }
  };
}
