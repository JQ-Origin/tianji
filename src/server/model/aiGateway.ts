import { buildQueryWithCache } from '../cache/index.js';
import { prisma } from './_client.js';
import { type RequestHandler } from 'express';
import { calcMessagesToken, calcOpenAIToken } from '../model/openai.js';
import { z } from 'zod';
import OpenAI from 'openai';
import { AIGatewayLogs, AIGatewayLogsStatus } from '@prisma/client';
import {
  getLLMCostDecimalV2,
  getLLMCostDecimalWithCustomPrice,
} from '../utils/llm.js';
import { get } from 'lodash-es';
import { verifyUserApiKey } from '../model/user.js';
import { checkQuotaAlert } from './aiGateway/quotaAlert.js';
import { logger } from '../utils/logger.js';

export const { get: getGatewayInfoCache, del: clearGatewayInfoCache } =
  buildQueryWithCache(async (workspaceId: string, gatewayId: string) => {
    const gatewayInfo = await prisma.aIGateway.findUnique({
      where: {
        workspaceId,
        id: gatewayId,
      },
    });

    return gatewayInfo;
  });

const openaiRequestSchema = z
  .object({
    model: z.string(),
    messages: z.array(z.any()).nonempty(),
    temperature: z.number().optional(),
    top_p: z.number().optional(),
    max_tokens: z.number().int().optional(),
    stream: z.boolean().optional(),
    presence_penalty: z.number().optional(),
    frequency_penalty: z.number().optional(),
  })
  .passthrough();

interface OpenaiHandlerOptions {
  baseUrl?: string;
  modelProvider?: string;
  modelPriceName?: (model: string) => string;
  isCustomRoute?: boolean;
}

export function buildOpenAIHandler(
  options: OpenaiHandlerOptions
): RequestHandler {
  return async (req, res) => {
    const payload = openaiRequestSchema.parse(req.body);
    const { messages, stream } = payload;
    const { workspaceId, gatewayId } = z
      .object({
        workspaceId: z.string(),
        gatewayId: z.string(),
      })
      .parse(req.params);
    const apiKey = (req.headers.authorization ?? '').replace('Bearer ', '');
    let modelApiKey = apiKey;
    const gatewayInfo = await getGatewayInfoCache(workspaceId, gatewayId);
    let userId: string | null = null;
    if (gatewayInfo?.modelApiKey) {
      const user = await verifyUserApiKey(apiKey);
      userId = user.id;
      modelApiKey = gatewayInfo.modelApiKey;
    }

    // Use custom settings from gateway if available
    const baseUrl =
      options.isCustomRoute && gatewayInfo?.customModelBaseUrl
        ? gatewayInfo?.customModelBaseUrl
        : options.baseUrl;
    const modelName =
      options.isCustomRoute && gatewayInfo?.customModelName
        ? gatewayInfo?.customModelName
        : payload.model;

    const start = Date.now();

    const logP = new Promise<AIGatewayLogs>(async (resolve) => {
      const _log = await prisma.aIGatewayLogs.create({
        data: {
          workspaceId,
          gatewayId,
          modelName,
          stream,
          inputToken: -1,
          outputToken: -1,
          duration: -1,
          ttft: -1,
          requestPayload: payload,
          responsePayload: {},
          userId,
          status: AIGatewayLogsStatus.Pending,
        },
      });

      resolve(_log);
    });

    try {
      const openai = new OpenAI({
        apiKey: modelApiKey,
        baseURL: baseUrl,
      });
      const modelProvider = options.modelProvider ?? 'openai';
      const modelPriceName = options.modelPriceName
        ? options.modelPriceName(modelName)
        : modelName;

      // Create payload with custom model name if specified
      const requestPayload = {
        ...payload,
        model: modelName,
      };

      // Handle stream response
      if (stream) {
        res.setHeader('Content-Type', 'text/event-stream');
        res.setHeader('Cache-Control', 'no-cache');
        res.setHeader('Connection', 'keep-alive');

        const stream = await openai.chat.completions.create({
          ...requestPayload,
          stream: true,
        });

        let outputContent = '';
        let ttft = -1;
        let usage: OpenAI.Completions.CompletionUsage = {
          prompt_tokens: 0,
          completion_tokens: 0,
          total_tokens: 0,
        };
        let responseModelName = modelName; // real model name which returned from remote. its will be some changed because of the model alias.
        for await (const chunk of stream) {
          responseModelName = chunk.model;
          if (ttft === -1) {
            ttft = Date.now() - start;
          }
          const content = get(chunk, ['choices', 0, 'delta', 'content']) || '';
          outputContent += content;
          res.write(`data: ${JSON.stringify(chunk)}\n\n`);

          if (chunk.usage) {
            usage = chunk.usage; // try to get usage from chunk
          }

          if (res.flush) {
            res.flush();
          }
        }

        res.write('data: [DONE]\n\n');
        res.end();
        const duration = Date.now() - start;

        logP.then(async ({ id: logId }) => {
          const inputToken =
            usage.prompt_tokens ??
            (await calcMessagesToken(messages, modelName));

          const outputToken =
            usage.completion_tokens ??
            (await calcOpenAIToken(outputContent, modelName));

          // Use custom price if available, otherwise use default pricing
          const customInputPrice = gatewayInfo?.customModelInputPrice;
          const customOutputPrice = gatewayInfo?.customModelOutputPrice;
          const price =
            options.isCustomRoute && (customInputPrice || customOutputPrice)
              ? getLLMCostDecimalWithCustomPrice(
                  inputToken,
                  outputToken,
                  customInputPrice,
                  customOutputPrice
                )
              : getLLMCostDecimalV2(
                  modelProvider,
                  modelPriceName,
                  inputToken,
                  outputToken
                );

          await prisma.aIGatewayLogs.update({
            where: {
              id: logId,
            },
            data: {
              status: AIGatewayLogsStatus.Success,
              modelName: responseModelName,
              inputToken,
              outputToken,
              duration,
              ttft,
              price,
              responsePayload: { content: outputContent, usage },
            },
          });

          // Check quota alert after successful request
          checkQuotaAlert(workspaceId, gatewayId, Number(price)).catch(
            (error) => {
              logger.error('Error checking quota alert:', error);
            }
          );
        });
      } else {
        // Handle normal response
        const response = await openai.chat.completions.create({
          ...requestPayload,
          stream: false,
        });

        const responseModelName = response.model ?? modelName;

        res.json(response);
        const duration = Date.now() - start;

        logP.then(async ({ id: logId }) => {
          const content = get(response, ['choices', 0, 'message', 'content']);

          const inputToken: number =
            response.usage?.prompt_tokens ??
            (await calcMessagesToken(messages, modelName));

          const outputToken =
            response.usage?.completion_tokens ??
            (typeof content === 'string'
              ? await calcOpenAIToken(content, modelName)
              : 0);

          // Use custom price if available, otherwise use default pricing
          const customInputPrice = gatewayInfo?.customModelInputPrice;
          const customOutputPrice = gatewayInfo?.customModelOutputPrice;
          const price =
            options.isCustomRoute && (customInputPrice || customOutputPrice)
              ? getLLMCostDecimalWithCustomPrice(
                  inputToken,
                  outputToken,
                  customInputPrice,
                  customOutputPrice
                )
              : getLLMCostDecimalV2(
                  modelProvider,
                  modelPriceName,
                  inputToken,
                  outputToken
                );

          await prisma.aIGatewayLogs.update({
            where: {
              id: logId,
            },
            data: {
              status: AIGatewayLogsStatus.Success,
              inputToken,
              outputToken,
              duration,
              modelName: responseModelName,
              price,
              responsePayload: { ...response },
            },
          });

          // Check quota alert after successful request
          checkQuotaAlert(workspaceId, gatewayId, Number(price)).catch(
            (error) => {
              logger.error('Error checking quota alert:', error);
            }
          );
        });
      }
    } catch (error) {
      // Handle API error
      console.error('OpenAI API error:', error);
      res.status(500).json({
        error: {
          message: error instanceof Error ? error.message : 'Unknown error',
          type: 'server_error',
        },
      });

      const duration = Date.now() - start;
      logP.then(async ({ id: logId }) => {
        await prisma.aIGatewayLogs.update({
          where: {
            id: logId,
          },
          data: {
            status: AIGatewayLogsStatus.Failed,
            duration,
            responsePayload: {
              error: String(error),
            },
          },
        });
      });
    }
  };
}
