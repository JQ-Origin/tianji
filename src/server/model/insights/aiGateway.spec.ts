import { describe, expect, test } from 'vitest';
import { AIGatewayInsightsSqlBuilder } from './aiGateway.js';
import { unwrapSQL } from '../../utils/prisma.js';

describe('AIGatewayInsightsSqlBuilder', () => {
  const insightId = 'cm95r56tv001gqhzlrxeyi6jm';
  const insightType = 'aigateway';

  test('basic query count', () => {
    const builder = new AIGatewayInsightsSqlBuilder(
      {
        insightId,
        insightType,
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1744273003917,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('basic query count');
  });

  test('basic query inputToken', () => {
    const builder = new AIGatewayInsightsSqlBuilder(
      {
        insightId,
        insightType,
        metrics: [
          {
            name: 'inputToken',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1744273003917,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('basic query inputToken');
  });
});
