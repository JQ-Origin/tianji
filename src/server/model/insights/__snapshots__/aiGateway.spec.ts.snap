// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AIGatewayInsightsSqlBuilder > basic query count > basic query count 1`] = `
"select
      to_char(date_trunc('day', "AIGatewayLogs"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date,
      count(1) as "$all_event"
    from "AIGatewayLogs" 
    where "AIGatewayLogs"."gatewayId" = 'cm95r56tv001gqhzlrxeyi6jm' AND "AIGatewayLogs"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-04-10T08:16:43.917Z'::timestamptz
    group by 1"
`;

exports[`AIGatewayInsightsSqlBuilder > basic query inputToken > basic query inputToken 1`] = `
"select
      to_char(date_trunc('day', "AIGatewayLogs"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date,
      sum("AIGatewayLogs"."inputToken") as "inputToken"
    from "AIGatewayLogs" 
    where "AIGatewayLogs"."gatewayId" = 'cm95r56tv001gqhzlrxeyi6jm' AND "AIGatewayLogs"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-04-10T08:16:43.917Z'::timestamptz
    group by 1"
`;
